File: 583-delete-operation-for-two-strings/delete-operation-for-two-strings.java
class Solution {
    public int minDistance(String word1, String word2) {
        return Min_OpsBU(word1,word2);
        
    }
    public static int Min_Ops(String s,String t,int i,int j){
        if(i==s.length()){
            return t.length()-j;
        }
        if(j==t.length()){
            return s.length()-i;
        }

        int ans=0;
        if(s.charAt(i)==t.charAt(j)){
            ans=Min_Ops(s,t,i+1,j+1);
        }
        else{
            int d1=Min_Ops(s,t,i+1,j);
            int d2=Min_Ops(s,t,i,j+1);
            ans=Math.min(d1,d2)+1;
        }
        return ans;
    }
    public int Min_OpsBU(String s,String t){
        int[][] dp=new int[s.length()+1][t.length()+1];
        for(int i=1;i<dp.length;i++){
            dp[i][0]=i;
        }
        for(int i=1;i<dp[0].length;i++){
            dp[0][i]=i;
        }
        for(int i=1;i<dp.length;i++){
            for(int j=1;j<dp[0].length;j++){
                if(s.charAt(i-1)==t.charAt(j-1)){
                    dp[i][j] = dp[i - 1][j - 1];
                }
                else{
                    int d1=dp[i-1][j];
                    int d2=dp[i][j-1];
                    dp[i][j]=Math.min(d1,d2)+1;
                }
            }
        }
        return dp[dp.length-1][dp[0].length-1];
    }
}

File: 583-delete-operation-for-two-strings/delete-operation-for-two-strings.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>> dp(word1.length(),vector<int>(word2.length(),-1));
        return dis(word1,word2,0,0,dp);
    }
    int dis(string s, string t,int i,int j,vector<vector<int>> &dp){
        if(i==s.length()){
            return t.length()-j;
        }
        if(j==t.length()){
            return s.length()-i;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int ans=0;
        if(s[i]==t[j]){
            ans=dis(s,t,i+1,j+1,dp);
        }
        else{
            int l=dis(s,t,i+1,j,dp);
            int d=dis(s,t,i,j+1,dp);
            ans=min(l,d)+1;
        }
        return dp[i][j]=ans;
    }
};